src/openassetio-python/tests/bridge/python/test_converter.cpp

```diff
@@ -0,0 +1,305 @@
+        PyObject* resultBool = PyObject_CallMethod(pyTraitsData, "hasTrait", "s", kTestTraitId);
+
+        // Use CHECK rather than REQUIRE in order to ensure the Decrefs
+        // below actually execute.
+        CHECK(PyBool_Check(resultBool));
+        CHECK(PyObject_IsTrue(resultBool));
+        Py_DECREF(resultBool);
+      }
+    }
+
+    WHEN("data is set via the Python object") {
+      PyObject_CallMethod(pyTraitsData, "addTrait", "s", kTestTraitId);
+
+      THEN("C++ object reflects the data set") { CHECK(traitsData->hasTrait(kTestTraitId)); }
+    }
+    Py_DECREF(pyTraitsData);
+  }
+
+  GIVEN("A Python object casted to a C++ object") {
+    // Use pybind to conveniently create a released CPython object.
+    const py::object pyClass = py::module_::import("openassetio").attr("TraitsData");
+    py::object pyInstance = pyClass();
+    PyObject* pyTraitsData = pyInstance.release().ptr();
+    const TraitsDataPtr traitsData =
+        openassetio::python::converter::castFromPyObject<TraitsData>(pyTraitsData);
+
+    WHEN("data is set via the C++ object") {
+      traitsData->addTrait(kTestTraitId);
+
+      THEN("Python object reflects that data set") {
+        PyObject* resultBool = PyObject_CallMethod(pyTraitsData, "hasTrait", "s", kTestTraitId);
+
+        // Use CHECK rather than REQUIRE in order to ensure the Decrefs
+        // below actually execute.
+        CHECK(PyBool_Check(resultBool));
+        CHECK(PyObject_IsTrue(resultBool));
+        Py_DECREF(resultBool);
+      }
+    }
+
+    WHEN("data is set via the Python object") {
+      PyObject_CallMethod(pyTraitsData, "addTrait", "s", kTestTraitId);
+      THEN("C++ object reflects the data set") { CHECK(traitsData->hasTrait(kTestTraitId)); }
+    }
+    Py_DECREF(pyTraitsData);
+  }
+}
+
+SCENARIO("Casting to PyObject extends object lifetime") {
+  GIVEN("A Python object casted from a C++ object") {
+    TraitsDataPtr traitsData = TraitsData::make();
+    traitsData->addTrait(kTestTraitId);
+    PyObject* pyTraitsData = openassetio::python::converter::castToPyObject(traitsData);
+
+    WHEN("C++ reference is destroyed") {
+      CHECK(Py_REFCNT(pyTraitsData) == 1);  // Initial condition.
+      traitsData.reset();
+
+      THEN("object remains alive and can be operated on via the Python interpreter") {
+        CHECK(Py_REFCNT(pyTraitsData) == 1);
+        PyObject* resultBool = PyObject_CallMethod(pyTraitsData, "hasTrait", "s", kTestTraitId);
+
+        // Use CHECK rather than REQUIRE in order to ensure the Decrefs
+        // below actually execute.
+        CHECK(PyBool_Check(resultBool));
+        CHECK(PyObject_IsTrue(resultBool));
+        Py_DECREF(resultBool);
+      }
+    }
+    Py_DECREF(pyTraitsData);
+  }
+}
+
+SCENARIO("Casting to a C++ object binds object lifetime") {
+  GIVEN("A python object") {
+    py::module_::import("openassetio");
+
+    // Use pybind to conveniently create a released CPython object.
```

Comment:
[consider]
```suggestion
    // Use pybind to conveniently create a CPython object with a ref count of 1.
```
Sorry to be a total pedant, but is 'released' confusing here? To me that wound be one with no ref count. Whereas its deliberately making one with an inc ref. As I understood it, "release" is the (imho confusingly named) method of `py::object` rather than a CPython thing - that abandons the underlying pyobject it previously increfd, which forms a 'trick' to give us a `PyObject` with a non-zero ref.

Reply:
acae406ef82b7f0342494d6bb37d19d12177ef07

Yeah, was just trying to convey how "pybind is done, CPython is king now". Have accepted your suggestion in the spirit of getting the thing across the line, it's more technically descriptive anyhow.

Reply:
🙏 thanks for the patience here!